### PR TITLE
Refactor coordinates

### PR DESCRIPTION
**Pull request by marscher**
* Discretizer is now a special case of Pipeline
* readers are now transformers (special case with self.data_producer = self).
* hide impl details with underscores in transformers
* cleanup api (hide impl detail with underscores)
* operate_in_memory refactored to setter of in_memory (properties). Ability to delete results

### CODE REVIEW

**Comment by franknoe** [comment]
Cool! Let me look through it tomorrow.

Am 30/03/15 um 19:50 schrieb Martin K. Scherer:
>
>   * Discretizer is now a special case of Pipeline
>   * readers are now transformers (special case with self.data_producer
>     = self).
>   * hide impl details with underscores in transformers
>   * cleanup api (hide impl detail with underscores)
>   * operate_in_memory refactored to setter of in_memory (properties).
>     Ability to delete results
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/markovmodel/PyEMMA/pull/170
>
>
>         Commit Summary
>
>   * [discretizer] base class is now Pipeline
>   * [transformer] made operate_in_memory a setter for in_memory
>   * [pipeline] fixes
>   * [readers] made them transformers, since they share 90% of the
>     interface
>   * [transformer] made map_to_memory private, since its called from
>     in_memory
>   * [mdtraj-patch] log critical message if traj is being loaded completly
>   * [setup] minor changes
>   * [datainmemory-test] wip
>   * [discretizer-test] use api functions
>   * [DataInMemory] reset time if itraj is incremented.
>   * removed dbg statements
>   * [coor/api] * discretizer does not use a default instance, since
>     this leads dramatic side effects.
>   * [coor/api] add featurizer to api
>   * [coor] hide impl detail in transformers (fixes #169)
>   * fix imports
>
>
>         File Changes
>
>   * *M* pyemma/coordinates/api.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-0> (72)
>   * *M* pyemma/coordinates/clustering/assign.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-1> (2)
>   * *M* pyemma/coordinates/clustering/kmeans.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-2> (21)
>   * *M* pyemma/coordinates/clustering/regspace.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-3> (10)
>   * *M* pyemma/coordinates/clustering/uniform_time.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-4> (14)
>   * *M* pyemma/coordinates/io/data_in_memory.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-5> (73)
>   * *M* pyemma/coordinates/io/feature_reader.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-6> (89)
>   * *D* pyemma/coordinates/io/reader.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-7> (23)
>   * *M* pyemma/coordinates/io/writer.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-8> (8)
>   * *R* pyemma/coordinates/pipeline.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-9> (278)
>   * *M* pyemma/coordinates/tests/test_datainmemory.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-10> (89)
>   * *M* pyemma/coordinates/tests/test_discretizer.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-11> (16)
>   * *M* pyemma/coordinates/tests/test_featurereader_and_tica.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-12> (2)
>   * *M*
>     pyemma/coordinates/tests/test_featurereader_and_tica_projection.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-13> (2)
>   * *M* pyemma/coordinates/tests/test_regspace.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-14> (4)
>   * *M* pyemma/coordinates/transform/pca.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-15> (10)
>   * *M* pyemma/coordinates/transform/tica.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-16> (34)
>   * *M* pyemma/coordinates/transform/transformer.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-17> (129)
>   * *M* pyemma/coordinates/util/chaining.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-18> (5)
>   * *M* pyemma/coordinates/util/patches.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-19> (1)
>   * *M* setup.py
>     <https://github.com/markovmodel/PyEMMA/pull/170/files#diff-20> (8)
>
>
>         Patch Links:
>
>   * https://github.com/markovmodel/PyEMMA/pull/170.patch
>   * https://github.com/markovmodel/PyEMMA/pull/170.diff
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/markovmodel/PyEMMA/pull/170>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by franknoe** [comment]
Looks ok to me. The only thing I have doubts about is the 
map_to_memory() and operate_in_memory() which should be possible to 
access from the outside, but I guess it doesn't necessarily need to be 
visible to the user as long as we provide some way of allowing them to 
do things in memory when they want (see API suggestion notebook)

-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by marscher** [comment]
If you set transformer.in_memory = True, _map_to_memory will be called internally, so you do not need to call it yourself.

File: pyemma/coordinates/io/data_in_memory.py [hunk 1]
```diff
@@ -130,37 +126,58 @@ def dimension(self):
...
+        # chunked mode
+        else:
+            upper_bound = min(self._t + self._chunksize, traj_len)
+            slice_x = slice(self._t, upper_bound)
+
+            X = traj[slice_x]
+            self._t += X.shape[0]
+
+            if lag == 0:
+                if self._t >= traj_len:
```
Comment (fabian-paul) [inline pyemma/coordinates/io/data_in_memory.py]: move out of lag if statement and delete special case, since you always iterate over all data.

**Comment by marscher** [comment]
since I've published this branch in this repo and work continues there - close it for now.

**Pull request closed without merging**